README.md
# SpaceBots

SpaceBots is a programming game where players employ their in-browser
javascript engines to drive hordes of robots and wage battles in
space.

## Running

After cloning complete dependencies:

 npm install

Before starting server, open two consoles to monitor state of the
server:

 touch debug.log && tail -f debug.log

 touch exceptions.log && tail -f exceptions.log

Now you are ready to roll:

 npm start

Server will start on port 8000 (if you are normal user) or on 443 (if
run under root).

Changelog
2013-05-17  Marek Rogalski  <mafikpl@gmail.com>

	ANNOUNCEMENT: This file will be used to notify users about changes
	in the game. Until there are any users, it does not need to be
	updated...

2013-05-16  Krzysztof H.  <krzys_h@interia.pl>

	* client.js, common.js (get): moved to common.js, to make it
	available in server console

2013-05-15  Marek Rogalski  <mafikpl@gmail.com>

	* index.html, client.js (impulse): renamed to do_impulse

	* client.js (get): added documentation

	* client.js (get_resources_for_waste): externalized from impulse

	* client.js (do_impulse, navigate): organized and extended
	documentation

	* client.js (speed_step): added momentum-aware velocity-changing
	function

	* client.js (stop_tick): switched implementation to speed_step,
	increased acceleration when stopping

2013-05-14  Krzysztof H.  <krzys_h@interia.pl>

	* client.js (stop): implemented stopping button

2013-05-13  Marek Rogalski  <mafikpl@gmail.com>

	* server.js (on 'report'): added 'manipulator_range' as a property
	valid for reporting.

	* client.js (on 'report'): added manipulator variable - that holds
	reference to last seen manipulator.

	* client.js (custom_scripts): added instructions for running local
	and unsecured scripts in chrome and chromium.

	* client.js (impulse): improved documentation, added two
	excercises.

	* client.js (navigate): pulled payload and imulse into separate
	variables.

	* client.js (intro, intro 'click' handler): added check for
	inhibitIntro that removes initial intro message. Added some
	documentation about localStorage.

	* client.js (createControls): renamed to controls

	* server.js (require 'ofe', tls.SLAB_BUFFER_SIZE): removed

	* LICENSE, WARRANTY: added license and warranty.

	* server.js: added console with license information.

	* resources.js (make_resources): added hidden inspect method to
	display resources.

	* server.js (attract, apply_secret_force): externalized encrypted
	apply_secret_force.


server.js

var	fs = require('fs');
var sty = require('sty');

var logger = require('./logger');

global.warranty = function() {
	process.stdout.write(fs.readFileSync('WARRANTY').toString());
};

global.copyright = function() {
	process.stdout.write(fs.readFileSync('LICENSE').toString());
}

var nesh = require('nesh');
nesh.log = logger;
nesh.config.load();
nesh.start({
	prompt: '> ',
	useGlobal: true,
	useColors: true,
	ignoreUndefined: true,
	historyFile: '.spacebots_history',
	historyMaxInputSize: 1024 * 1024,
	welcome: sty.b('SpaceBots')+' Copyright (C) 2013  Marek Rogalski\n' +
		'This program comes with ABSOLUTELY NO WARRANTY; for details type "warranty()".\n' + 
		'This is free software, and you are welcome to redistribute it under certain\n' +
		'conditions; type "copyright()" for details.'
}, function(err, repl) {
	if(err) {
		logger.error(err);
		return;
	}

	repl.on('exit', function () {
		process.stdout.write(sty.u('(keyboard exit)\n'));
		process.exit();
	});
});

var file = new (require('node-static').Server)('./static');

if( process.getuid() === 0 ) {
	app = require('https').createServer(
		{
			key: fs.readFileSync('privatekey.pem').toString(),
			cert: fs.readFileSync('certificate.pem').toString()
		}, 
		function (req, res) {
			file.serve(req, res);
		}
	);
	port = 443;

	redirector = require('http').createServer(function (req, res) {
		res.writeHead(302, {
			'Location': 'https://' + req.headers['host'] + '/'
		});
		res.end();
	});
	redirector.listen(80);
} else {
	app = require('http').createServer(function (req, res) {
		file.serve(req, res);
	});
	port = 8000;
}

var io = require('socket.io').listen(app, { logger: logger });
// io.set('log level', 1);

app.listen(port);
logger.info('Listening on port ' + port);

// Begin game logic

var common = require('./static/common'),
	sylvester = require('sylvester'),
	resources = require('./static/resources'),
	bp = require('./blueprints');

var objects = global.objects = {},
	Vector = sylvester.Vector;

var reg = function(obj) {
	objects[obj.id] = obj;
	return obj;
};

var place = function(host, guest, pos) {
	host.skeleton_slots[pos] = guest;
	guest.parent = host;
	return guest;
};

var asteroids = global.asteroids = [];

for(var i = 0; i < 30; ++i) {
	var asteroid = reg(resources.make_asteroid());
	asteroid.position = common.RV(3000);
	asteroid.velocity = common.RV(5);
	asteroids.push(asteroid);
}

var get_or_create_player = function(hash) {
    if(!(hash in objects)) {

		var hull = reg(bp.make('skeleton manipulator', 10));
		hull.position = common.RV(200);
		hull.velocity = common.RV(5);
		hull.skeleton_slots = [null, null, null, null, null, null];
		hull.sprite = '/hull.png';

		var avatar = place(hull, reg(bp.make('avatar radar', 10)), 0);
		avatar.radar_range = 1000;

		var drive = place(hull, reg(bp.make('impulse_drive store battery', 10)), 1);
		drive.store_stored[0] = drive.store_capacity;
		drive.battery_energy = drive.battery_capacity;
		place(hull, reg(bp.make('assembler refinery spectrometer', 10)), 2);
		place(hull, reg(bp.make('laboratory enriching_reactor burning_reactor', 10)), 3);

		var player = reg({
			id: hash,
			avatars: {}, 
			last_commands: [0,0,0,0,0,0,0,0,0,0]
		});

		player.avatars[avatar.id] = avatar;
    }
    return objects[hash];
};

var colors = 'red green yellow blue magenta cyan'.split(' ');

var random_color = function() {
	var i = Math.floor(Math.random() * colors.length);
	return colors[i];
};

var stub = function(obj) {
	if(obj) return { id: obj.id };
};

attractor = reg(resources.make_asteroid());
attractor.sprite = '/attractor151.png';

apply_secret_force = function(object) {
	for(var i = 1; i <= 3; ++i) {
		var v = object.position.e(i);
		if(v < -2000) {
			object.position.elements[i-1] = - v - 4000;
			object.velocity.elements[i-1] = - object.velocity.elements[i-1];
		}
		if(object.position.e(i) > 2000) {
			object.position.elements[i-1] = - v + 4000;
			object.velocity.elements[i-1] = - object.velocity.elements[i-1];
		}
	}
};

if('ATTRACTOR_KEY' in process.env) {
	try {
		var decipher = require('crypto').
			createDecipher('aes256', process.env.ATTRACTOR_KEY);
		decipher.end('k7S4ZzSGD7L4v7rOZXAiHgNJHbBGSbAUzwb+EWEjYau5//l0Rep1Hq' +
					 'hZFDI7qKOH6gqmsyr+plzaqL+1MNrHkzepOcMVkeCjWbwIU+xF19g=',
					 'base64');
		var str = decipher.read().toString('utf8');
		apply_secret_force = eval(str);
		logger.info('attractor\'s secret decrypted');
	} catch(e) {
		logger.error('couldn\'t decrypt attractor secret: ' + e.toString());
	}
}

attract = function() {
	for(var hash in objects) {
		var object = objects[hash];
		if(object.position && object.velocity) {
			object.position = object.position.add(object.velocity.x(0.1));
			apply_secret_force(object);
		}
	}
	setTimeout(attract, 100);
};

attract();

io.sockets.on('connection', function (socket) {
    var player = undefined;
	var address = socket.handshake.address;

	// socket.id

	var name = sty.b(sty[random_color()](address.address));
	var current_handler = 'unknown';

	(function() {
		var old_emit = socket.emit;
		socket.emit = function() {
			logger.info(name + ' < ' + arguments[0]);
			old_emit.apply(socket, arguments);
		};
	})();

	var log_in = function(message) {
		current_handler = message;
		logger.info(name + ' > ' + message);
	};

	var log = function(message) {
		logger.info(name + ' : ' + message);
	};

	var fail = function(code, message) {
		socket.emit('fail', {
			code: code,
			source: current_handler,
			message: message
		});
	};

	log('connected');

	
	socket.on('broadcast', function(data) {
		log_in('broadcast');
		var str = JSON.stringify(data);
		if(str.length > 140) {
			return fail(1, 'Message json should have at most 140 characters');
		}
		socket.broadcast.emit('broadcast', data);
	});

	socket.on('log in', function(data) {
		log_in('log in');
		if(!('' + data.player_id).match(/[0-9A-F]{32}/i)) {
			return fail(2, 'Hash used to log in does not match regular' +
						' expression /[0-9A-F]{32}/i .');
		}
		player = get_or_create_player('' + data.player_id);
		if(!('avatars' in player)) {
			player = undefined;
			return fail(3, 'This account doesn\'t have an avatar list');
		}
		socket.emit('avatar list',  Object.keys(player.avatars));
	});

	var find_target = function(command) {
		if(typeof player === 'undefined') {
			return fail(4, 'No player selected. You should first log in.');
		}
		if(typeof command.target === 'undefined') {
			return fail(5, 'Command didn\'t have `target` defined.');
		}
		if(!('' + command.target).match(/[0-9A-F]{32}/i)) {
			return fail(6, 'Target hash is not a valid identifier (should' +
						' match /[0-9A-F]{32}/i).');
		}
		// TODO: zasięg poleceń dla awatarów
		
		var visited = {};
		var queue = [];
		for(var id in player.avatars) {
			queue.push(id);
		}
		var count = 0;
		while(queue.length) {
			id = queue.pop();
			if(id in visited) continue;
			visited[id] = objects[id];
			count += 1;
			var current = objects[id];
			if(id == command.target) return current;
			if(current.parent) {
				queue.push(current.parent.id);
			}
			if(current.skeleton_slots) {
				for(var i = 0; i < current.skeleton_slots.length; ++i) {
					if(current.skeleton_slots[i]) {
						queue.push(current.skeleton_slots[i].id);
					}
				}
			}
		}
		return fail(7, 'Command target not found connected to any ' + 
				'avatar (searched ' + count+' objects).');
	};

	var check_feature = function(object, feature) {
		if(typeof object.features[feature] === 'undefined') {
			return fail(8, 'Specified component doesn\'t  have ' + 
					feature + ' capabilities');
		}
		return true;
	};

	var on = function(name, handler) {
		socket.on(name, function(cmd) {
			var now = (new Date).getTime() / 1000;
			if(typeof player === 'undefined') {
				return fail(18, 'You have to log in first!');
			}
			var ten_before = player.last_commands.shift();
			player.last_commands.push(now);
			if(now - ten_before < 1) {
				return fail(9, 'Exceeded limit of ' + player.last_commands.length + ' commands per second.');
			}
			log_in(name);
			var target = find_target(cmd);
			if(!target) return;
			handler(target, cmd);
		});
	};

	on('report', function(target, data) {
		var report = {};
		if('parent' in target) {
			report.parent = stub(target.parent);
		}
		if('skeleton_slots' in target) {
			report.skeleton_slots = target.skeleton_slots.map(stub);
		}
		var copy = 'id features sprite radar_range impulse_drive_payload impulse_drive_impulse store_stored store_capacity battery_energy battery_capacity manipulator_range'.split(' ');
		var vectors = 'position velocity'.split(' ');
		copy.forEach(function(key) {
			report[key] = target[key];
		});
		vectors.forEach(function(key) {
			if(key in target) {
				report[key] = target[key].elements;
			}
		});
		if(target.composition) {
			report.mass = resources.get_mass(target.composition);
		}
		socket.emit('report', report);
	});

	var radar_copy_fields = 'id sprite'.split(' ');
	var radar_vector_fields = 'position velocity'.split(' ');
	on('radar scan', function(target, data) {
		if(!check_feature(target, 'radar')) return;
		var results = [];
		var now = (new Date).getTime();
		var radar_position = common.get_position(target);
		
		for(var hash in objects) {
			var object = objects[hash];
			if(!('position' in object)) continue;
			var d = radar_position.distanceFrom(object.position);
			if(d <= target.radar_range) {
				var report = {};
				radar_copy_fields.forEach(function(key) { report[key] = object[key]; });
				radar_vector_fields.forEach(function(key) { report[key] = object[key].elements; });
				
				results.push(report);
			}
		}
		socket.emit('radar result', results);
	});

	var find_co_component = function(source, id, feature) {
		var cc = common.walk(source);
		if(!(id in cc)) {
			return fail(10, 'Object ' + id + ' is not reachable from ' +
						source.id);
		}
		if(!check_feature(cc[id], feature)) return undefined;
		return cc[id];
	};

	on('impulse_drive push', function(target, cmd) {
		if(!check_feature(target, 'impulse_drive')) return;

		var energy_source = find_co_component(target, cmd.energy_source, 'battery');
		if(typeof energy_source === 'undefined') return;
		var matter_store = find_co_component(target, cmd.matter_source, 'store');
		if(typeof matter_store === 'undefined') return;

		if(!resources.lte(cmd.composition,matter_store.composition)) {
			return fail(11, 'Ordered to grab more materials than available in store.');
		}
		var reaction_mass = resources.get_mass(cmd.composition);
		if(reaction_mass > target.impulse_drive_payload) {
			return fail(12, 'Ordered payload exceeds drive capabilities.');
		}
		if(cmd.impulse > target.impulse_drive_impulse) {
			return fail(13, 'Ordered power exceeds drive capabilities.');
		}
		var momentum = reaction_mass * cmd.impulse;

		if(momentum > energy_source.battery_energy) {
			return fail(14, 'Ordered to use more power than available in battery.');
		}
		
		var ship_mass = resources.get_connected_mass(target);
		
		// TODO: fajne testy na rozerwanie przez przyspieszenie

		// Warning: power === momentum!
		var delta_v = momentum / ship_mass;

		var root = common.get_root(target);

		var trajectory = Vector.create(cmd.destination).subtract(root.position);

		var time = trajectory.modulus() / cmd.impulse;
		setTimeout(function() {
			socket.emit('explosion', {
				sprite: '/explosion45.png',
				duration: 1,
				position: cmd.destination
			});

			var d = trajectory.modulus();
			var r = d * reaction_mass * cmd.impulse / 100000;
			var dmg = momentum;
			var hit_point = Vector.create(cmd.destination);
			var noluck = {};
			var total = 0;
			var hit = [];

			for(var hash in objects) {
				var o = objects[hash];
				if(o.position) {
					var l = o.position.distanceFrom(hit_point);
					if(l < r) {
						var m = resources.get_connected_mass(o);
						total += m / l;
						hit.push(o);
					}
				}
			}
		}, time * 1000);

		trajectory = trajectory.toUnitVector().x(-delta_v);
		root.velocity = root.velocity.add(trajectory);

		resources.subtract(matter_store.store_stored, cmd.composition);
		energy_source.battery_energy -= momentum;

		/*
		console.log("Matter ejected: ", reaction_mass);
		console.log("Impulse: ", cmd.impulse);
		console.log("Momentum: ", momentum);

		console.log("Ship mass: ", ship_mass);
		console.log("Change of velocity: ", delta_v);

		console.log("Ship position", root.position.elements);
		console.log("Target position", cmd.destination);
		*/
		
	});

	(function() {

		var random_asteroid_id = function() {
			var i = Math.floor(Math.random() * asteroids.length);
			return asteroids[i].id;
		};

		var challenge;

		var send_challlenge = function() {
			if(socket.connected) {
				challenge = random_asteroid_id();
				socket.emit('location challenge', challenge);
				setTimeout(send_challlenge, 1000);
			}
		};

		socket.on('location answer', function(ans) {
			log_in('location answer');
			if(ans.id !== challenge) {
				return fail(15, 'Answer to wrong challenge!');
			}
			var p = Vector.create(ans.position);
			var v = Vector.create(ans.velocity);

			var o = objects[challenge];
			challenge = undefined;

			if(typeof o === 'undefined') {
				return fail(16, 'You don\'t have any challenge');
			}

			var op = o.position;
			var ov = o.velocity;

			var dp = op.distanceFrom(p);
			var dv = ov.distanceFrom(v);

			var dist = 99999;
			for(hash in player.avatars) {
				var ap = common.get_position(player.avatars[hash]);
				dist = Math.min(dist, ap.distanceFrom(op));
			}

			// Otworzyć plik odpowiadający temu IP, dopisać do niego
			// dist dp dv
			fs.appendFile('./logs/' + address.address + '.log',
						  [(new Date).getTime(), dist, dp, dv].join(' ') + '\n',
						  function (err) {
							  if(err) {
								  fail(17, 'Couldn\'t append to log file on server. Call the developers...');
							  }
						  });
			
		});

		send_challlenge();

	})();
});
